### DISCUSSION TITLE
Fastest way to get the MauiContext? Is MauiContext same for all objects in a project? Is it stable over the app's runtime? What does this do?

### DISCUSSION THREAD

**Discussion by jonmdev** [dotnet/maui#18364] [General]
Many of the `ToPlatform()` type methods require a MauiContext. I see that objects don't have MauiContext set until they have their Handlers assigned. Handlers are assigned once the native objects have been set up and added to the hierarchy and are thus ready for Maui usage. 

Am I correct so far?

I think MauiContext is just some reference that is set through all Maui Handlers to let them talk back to the Maui App or shared other infrastructure. Ie. I think the MauiContext reference is the same for all objects always. 

Is that correct too?

I see this summary which gives some clues:

>Typically, the pattern for initializing .NET MAUI in a native app project is to:
>
>- create a xref:Microsoft.Maui.Hosting.MauiAppBuilder object, 
>- invoke the xref:Microsoft.Maui.Embedding.AppHostBuilderExtensions.UseMauiEmbedding%2A method on it, 
>- and then create a xref:Microsoft.Maui.Hosting.MauiApp object by invoking the xref:Microsoft.Maui.Hosting.MauiAppBuilder.Build method on the xref:Microsoft.Maui.Hosting.MauiAppBuilder object. 
> 
>In addition, a xref:Microsoft.Maui.MauiContext object should then be created from the xref:Microsoft.Maui.Hosting.MauiApp object. 
>
>The xref:Microsoft.Maui.MauiContext object will be used when converting .NET MAUI controls to native types.

https://github.com/dotnet/docs-maui/blob/main/docs/platform-integration/native-embedding.md

If so, and we know we have numerous functions needing `MauiContext`, what is the fastest way to get and store the reference to MauiContext (for example in a Singleton Instance)?

My suspicion is the correct way to do this then is:

```
public partial class App : Application {

    public App() {
         this.HandlerChanged+=delegate{
              saveGlobalMauiContext(this.Handler.MauiContext);
}
```

I presume the App must get its handler first and thus represents the first likely access to the MauiContext object? Is this correct or am I understanding this correctly?

Lastly, I have noted Handlers can change several times for some objects on startup (not sure why, haven't looked in detail and which ones off the top of my head), but is MauiContext expected to do the same?

Ie. Does MauiContext change over the App's lifespan? If we save a reference on startup is it good until the app restarts?

Thanks.

**Comment by JTOne123** [comment]
here is another example https://github.com/Redth/FFImageLoading.Compat/blob/main/source/FFImageLoading.Maui/MauiExtensions.cs

**Comment by jingliancui** [comment]
That's my way:
```
var appHandler = Application.Current.Handler;
var xxService ??= appHandler!.MauiContext!.Services.GetRequiredService<MyService>();
```
Specially the case in page's ctor, if using this.Handler or base.Handler, all will return null, so I try "Application" handler, any ideas from your perspective? @mattleibow @jfversluis @PureWeen
